File: 508. Most Frequent Subtree Sum/solution.cpp
/*
19 ms
24.6 MB
*/
/**
 * TC: O(n)
 * SC: O(height)
*/
/**
 * @author : SahilK-027
 * @brief : DFS
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:    
    int solve(TreeNode* root,unordered_map<int, int>& mp){
        // Base case
        if(!root){
            return 0;
        }

        // Solve left and right subtree recursively
        int left_ans = solve(root->left, mp);
        int right_ans = solve(root->right, mp);

        // Find sum at root
        int sum = left_ans + right_ans + root->val;

        // Mark entry in map ans return sum
        mp[sum]++;
        return sum;
    }
public:
    vector<int> findFrequentTreeSum(TreeNode* root) {
        // We will use map to store sum and its count
        unordered_map<int, int>mp;
        vector<int> ans;
        solve(root, mp);

        // Finding the max sum from map
        int max_cnt = INT_MIN;
        for(auto i: mp){
            if(i.second > max_cnt){
                max_cnt = i.second;
            }
        }

        // Adding nodes whose sum is max sum to answer
        for(auto i: mp){
            if(i.second == max_cnt){
                ans.push_back(i.first);
            }
        }
        return ans;
    }
};

File: 508. Most Frequent Subtree Sum/solution.js
/*
85 ms
47.1 MB
*/
/**
 * TC: O(n)
 * SC: O(height)
*/
/**
 * @author : SahilK-027
 * @brief : DFS
 */ 

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

function solve(root, mp){
    if(!root){
        return 0;
    }

    let left_ans = solve(root.left, mp);
    let right_ans = solve(root.right, mp);

    let sum = root.val + left_ans+ right_ans;

    mp[sum] = (mp[sum] || 0) + 1;
    return sum;
}
var findFrequentTreeSum = function(root) {
    // Declaring map
    let mp = {};
    let max = -Infinity;
    let ans = [];

    // Recursive call
    solve(root, mp);

    // Finding Most Frequent from map
    for(let key in mp){
        if(mp[key] > max){
            max = mp[key];
        }
    }

    // Storing Most Frequent elem in ans
    for(let key in mp){
        if(mp[key] === max){
            ans.push(parseInt(key));
        }
    }
    return ans;
};